sdks/python/apache_beam/transforms/combiners.py

```diff
@@ -181,8 +184,22 @@ def Of(pcoll, n, compare=None, *args, **kwargs):
     """
     key = kwargs.pop('key', None)
     reverse = kwargs.pop('reverse', False)
-    return pcoll | core.CombineGlobally(
-        TopCombineFn(n, compare, key, reverse), *args, **kwargs)
+    if not args and not kwargs and not key and pcoll.windowing.is_default():
+      if reverse:
+        if compare is None or compare is operator.lt:
+          compare = operator.gt
+        else:
+          original_compare = compare
+          compare = lambda a, b: original_compare(b, a)
+      # This is a more efficient global algorithm.
+      return (
```

Comment:
This implementation seems to prevent the equivalent of "combiner lifting" for a future where one might take advantage of Multi Shard Combining. Could that be problematic?

Reply:
That is correct. There are two issues here that lead to this structure: 

1) I want to accumulate into a heap, but then sort the accumulator before emitting it to shift work from the (single) reducer to the (many) mappers. It's unclear how to express this as a CombineFn. Perhaps https://issues.apache.org/jira/browse/BEAM-4030 could help (if we ensure that it is called on all runners). 
2) I want to avoid encoding _ComparableValue objects, which will be done via pickling. This could possibly be worked around if we could set and enforce custom coders. 

Perhaps a single accumulator object with a custom __reduce__ would suffice.